### PR TITLE
Added: Templates inheritance

### PR DESCRIPTION
**Pull request by simobasso**
Hey @cookiecutter,
This implements a templates directory so a maintainer can use `extends`, `includes`, `blocks`, `import` and `super` inside the cookiecutter project template using jinja.

loader: https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.FileSystemLoader
template-inheritance docs: https://jinja.palletsprojects.com/en/2.11.x/templates/#template-inheritance

The ratio is described in:

closes: https://github.com/cookiecutter/cookiecutter/issues/1484

other issues on the same topic:

closes: https://github.com/cookiecutter/cookiecutter/issues/1495
closes: https://github.com/cookiecutter/cookiecutter/issues/818
refs: https://github.com/cookiecutter/cookiecutter/issues/409

similar/extendable issues that get an improvement from this pr:

ref: https://github.com/cookiecutter/cookiecutter/issues/706
ref: https://github.com/cookiecutter/cookiecutter/issues/59
ref: https://github.com/cookiecutter/cookiecutter/issues/1004
ref: https://github.com/cookiecutter/cookiecutter/issues/364

Thank you!

### CODE REVIEW

File: cookiecutter/generate.py
Comment (simobasso) [inline cookiecutter/generate.py:309]: as reference https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.FileSystemLoader

**Comment by simobasso** [comment]
@cookiecutter friendly ping 😊

**Comment by sisp** [comment]
@cookiecutter Any change of getting this PR merged? It's exactly what I've been looking for. Thanks @simobasso for implementing it!

**Comment by simobasso** [comment]
Hey @audreyr @pydanny @hackebrot, any chance of getting this pr reviewed?

**Comment by pydanny** [comment]
Apologies, but none of us have the unpaid bandwidth to work on this project. 

I think we need to do another call for maintainers.

**Comment by simobasso** [comment]
@pydanny, thanks anyway!

**Comment by HappyEinara** [comment]
Hi all, I'd like to +1 this proposal because I think it would unlock a very powerful application for cookiecutter: composability. For example, if I want a template to provide my application with an API on AWS Lambda + API Gateway, then I could include subtemplates "API" (providing FastAPI and some endpoint stubs and tests) and "AWS Lambda API" implementing the infra. That in turn could inherit from an AWS base template and be a sibling to an API-on-ECS template.

While some of the more generic bits could be done by using separate python packages as the engine for composition, I think generating optional application components (CLI, web API, cloud provider, Slack integration) to be included by flags in cookiecutter.json would be really, really powerful when used creatively and carefully, although I find Jinja does get much less readable once you move from variable interpolation to control flow and blocks.

I'm currently managing this sort of scenario using git-fu, but it takes a lot of attention to detail.

I wonder if @timothycrosley or @samj1912 from [Cruft](https://github.com/cruft/cruft) would have the skills, interest or time to help with this?

I do have one question on the implementation: the PR seems well able to include blocks of file content, but just as important would be the optional inclusion of whole files, directories or trees of directories in the same way (and analogously to the way cookiecutter permits templating in filenames). @simobasso do you have any thoughts on how that would be done?

**Comment by samj1912** [comment]
@HappyEinara thanks for the call-out!

@pydanny I would be more than happy to help out with the maintanance of cookiecutter (I currently maintain http://cruft.github.io/ and https://github.com/samj1912/cookies which heavily rely on cookiecutter internals so I am fairly familiar with the codebase) but I would imagine we would still need more people (at least 1 more active maintainer) to be able to go through the current backlog or PRs and issues.

**Comment by simobasso** [comment]
@pydanny @samj1912 if this project needs some love, I'm more than happy to give it some :).

> I do have one question on the implementation: the PR seems well able to include blocks of file content, but just as important would be the optional inclusion of whole files, directories or trees of directories in the same way (and analogously to the way cookiecutter permits templating in filenames). @simobasso do you have any thoughts on how that would be done?

That's an interesting question. I will open a new issue to keep this one on track because I think it is a bit far away from the scope of this pr; I will contribute there.

**Review by samj1912 — APPROVED** [review]

File: cookiecutter/generate.py
Comment (samj1912) [inline cookiecutter/generate.py:309]: This should use os.path.sep instead so that it is compatible with windows and unix systems.

**Comment by sisp** [comment]
@cookiecutter Any chance for this PR to get merged? Or are there any remaining concerns? It's a really great feature I'd love to see become available. :slightly_smiling_face:

**Comment by tall-josh** [comment]
Hi. Firstly LOVE Cookiecutter!!! If I could :star:  more than once I would!!!

Is there any movement on this PR? I would really love this feature.

**Comment by JWBWork** [comment]
Any status update on this?

**Review by JWBWork — APPROVED** [review]
Looks good to me

**Comment by insspb** [comment]
Rebased to master

File: docs/advanced/templates.rst
Comment (insspb) [inline docs/advanced/templates.rst:3]:
```suggestion
Templates inheritance (2.2+)
```

File: docs/advanced/templates.rst
Comment (insspb) [inline docs/advanced/templates.rst:6]:
```suggestion
*New in Cookiecutter 2.2+*
```

**Review by insspb — APPROVED** [review]

**Comment by insspb** [comment]
@simobasso thank you for contribution and waiting. I will merge it, if you have a time please look at #1694

@sisp, @JWBWork, @tall-josh, @samj1912 this will be in next release

**Pull request merged**
